drivers: sensor: mcp9600 add initial driver support

maxmclau opened the pull request:
This driver adds support for the MCP9600 I2C Thermocouple Amplifier. It's mostly recycled from abandoned #31082 w/ fixes for proper multi-device support and `_dt` devicetree bindings.

Signed-off-by: Max McLaughlin [github@maxmclau.com](mailto:github@maxmclau.com)

galak commented on drivers/sensor/mcp9600/Kconfig:
```diff
@@ -0,0 +1,13 @@
+
+menuconfig MCP9600
+	bool "MCP9600 i2c thermocouple temperature sensor"
+	depends on I2C
```
Add:

```
default y
depends on DT_HAS_MICROCHIP_MCP9000_ENABLED
```

galak commented on drivers/sensor/mcp9600/mcp9600.c:
```diff
@@ -0,0 +1,97 @@
+  }
+
+  data->temp = (int32_t)(int16_t)(buf[0] << 8) | buf[1];
+  data->temp *= 62500;
```
Maybe add a comment about what this multiplication is doing.

galak reviewed (dismissed):
Update Kconfig to add support for auto-generated Kconfig.dts symbols.

galak commented:
@maxmclau be aware for final approve you will need to squash the commits down into one.

maxmclau commented:
> @maxmclau be aware for final approve you will need to squash the commits down into one.

squashed

maxmclau commented:
just checking - anything else missing here before merge?

MaureenHelm commented:
Please fix the compliance check failure

MaureenHelm commented:
The compliance fixes need to be amended into the original commit. See the [Contribution Workflow](https://docs.zephyrproject.org/latest/contribute/guidelines.html#contribution-workflow) for details.

galak commented on drivers/sensor/mcp9600/Kconfig:
```diff
@@ -0,0 +1,14 @@
+	bool "MCP9600 i2c thermocouple temperature sensor"
+	default y
+	depends on DT_HAS_MICROCHIP_MCP9600_ENABLED
+	depends on I2C
```
`depends on I2C` should now be `select I2C`

maxmclau replied:
should be resolved

MaureenHelm commented:
> > The compliance fixes need to be amended into the original commit. See the [Contribution Workflow](https://docs.zephyrproject.org/latest/contribute/guidelines.html#contribution-workflow) for details.
> 
> should be fixed now

Sorry, no. :( The commit message currently appears to be a list of subject lines without any body, and the signed-off-by line is mixed into the middle when it needs to be at the end. It should look something like this:
```
drivers: sensor: mcp9600 add intial driver support

A body that explains why you're making this change, not what the change
is. This can be multiple lines or paragraphs long, and lines need to
wrap at 75 columns.

Signed-off-by: Maxmillion McLaughlin <github@maxmclau.com>
```

Please look at the git log for examples.

napei commented:
Would love to see this merged, currently using this sensor.

kartben commented:
> Would love to see this merged, currently using this sensor.

+1, I would love that too as I am planning to use this sensor for a project as well!
@maxmclau do you need further assistance?

maxmclau commented:
@napei and @kartben - Sorry for the radio silence. Just cleaned up the commit and updated the driver to use `SENSOR_DEVICE_DT_INST_DEFINE`. Hopefully this should be clear to merge

teburd commented:
@teburd @MaureenHelm lets take a look at this

MaureenHelm commented on drivers/sensor/mcp9600/mcp9600.h:
```diff
@@ -0,0 +1,54 @@
+
+#define MCP9600_REG_TEMP_HOT 0x00
+
+#define MCP9600_REG_TEMP_DIFF 0x01
+#define MCP9600_REG_TEMP_COLD 0x02
+#define MCP9600_REG_RAW_ADC   0x03
+
+#define MCP9600_REG_STATUS 0x04
+
+#define MCP9600_REG_TC_CONFIG  0x05
+#define MCP9600_REG_DEV_CONFIG 0x06
+#define MCP9600_REG_A1_CONFIG  0x07
+
+#define MCP9600_REG_A2_CONFIG 0x08
+#define MCP9600_REG_A3_CONFIG 0x09
+#define MCP9600_REG_A4_CONFIG 0x0A
+#define MCP9600_A1_HYST	      0x0B
+
+#define MCP9600_A2_HYST	 0x0C
+#define MCP9600_A3_HYST	 0x0D
+#define MCP9600_A4_HYST	 0x0E
+#define MCP9600_A1_LIMIT 0x0F
+
+#define MCP9600_A2_LIMIT	0x10
+#define MCP9600_A3_LIMIT	0x11
+#define MCP9600_A4_LIMIT	0x12
+#define MCP9600_REG_ID_REVISION 0x13
+
+struct mcp9600_data {
+	int32_t temp;
+};
+
+struct mcp9600_config {
+	const struct i2c_dt_spec bus;
+};
```
This isn't really worth it's own private header file. Just move these definitions to mcp9600.c

MaureenHelm commented on dts/bindings/sensor/microchip,mcp9600.yaml:
```diff
@@ -0,0 +1,12 @@
+
+compatible: "microchip,mcp9600"
+
+include: i2c-device.yaml
```
```suggestion
include: [sensor-device.yaml, i2c-device.yaml]
```

MaureenHelm commented on tests/drivers/build_all/sensor/i2c.dtsi:
```diff
@@ -171,6 +171,11 @@ test_i2c_ms5837: ms5837@1a {
 	reg = <0x1a>;
 };
 
+test_i2c_mcp9600: mcp9600@60 {
+	compatible = "microchip,mcp9600";
+	reg = <0x60>;
+};
+
```
Have you rebased to the latest main recently? These addresses should be in numerical order

maxmclau replied:
Appended in correct numerical order

maxmclau committed:
drivers: sensor: mcp9600 add intial driver support

Implementation of MCP9600 i2c thermoouple amplifier - K, J, T, N, S, E, B
and R type T

Signed-off-by: Maxmillion McLaughlin <github@maxmclau.com>

maxmclau committed:
tests: drivers: build_all: sensor: mcp9600 test case

Adds test case for mcp9600 thermocouple amplifier.

Signed-off-by: Maxmillion McLaughlin <github@maxmclau.com>

maxmclau commented:
Sorry for the timeline mess! Review items have all been addressed

MaureenHelm commented on drivers/sensor/mcp9600/mcp9600.c:
```diff
@@ -0,0 +1,134 @@
+	.channel_get = mcp9600_channel_get,
+};
+
+int mcp9600_init(const struct device *dev)
```
```suggestion
static int mcp9600_init(const struct device *dev)
```

MaureenHelm commented on drivers/sensor/mcp9600/mcp9600.c:
```diff
@@ -0,0 +1,134 @@
+	uint8_t buf[2];
+	int ret;
+
+	if (!device_is_ready(cfg->bus.bus)) {
```
i2c_is_ready_dt

MaureenHelm reviewed (changes requested):
Needs rebase

maxmclau commented on drivers/sensor/mcp9600/mcp9600.c:
```diff
@@ -0,0 +1,134 @@
+{
+	struct mcp9600_data *data = dev->data;
+
+	__ASSERT_NO_MSG(chan == SENSOR_CHAN_AMBIENT_TEMP);
```
should I be returning an errno instead of using `__ASSERT_NO_MSG()`?

kartben replied:
Yes, `-ENOTSUP` is typically returned for unsupported channels. https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/sensor/iis2dh/iis2dh.c#L146-L163

maxmclau commented:
rebased and resolved two of three changes.

maxmclau commented:
resolved i2c test address issue and rebased. any other issues?

maxmclau commented:
Wow, I'm sorry figured I'd try the Github auto rebase to bring the branch up to date. Looks like it is requested review from nearly everyone. Branch [mcp9600-driver](https://github.com/maxmclau/zephyr/tree/mcp9600-driver) should be rebased and conflicts resolved, but feel free to close if it's a headache.

MaureenHelm commented on drivers/sensor/Kconfig:
```diff
@@ -193,6 +193,8 @@ source "drivers/sensor/mchp_tach_xec/Kconfig"
 
 source "drivers/sensor/mcp970x/Kconfig"
 
+source "drivers/sensor/mcp9600/Kconfig"
```
Move up to keep the numbers in order

MaureenHelm reviewed (changes requested):
One minor nitpick but otherwise LGTM

uLipe commented on drivers/sensor/mcp9600/mcp9600.c:
```diff
@@ -0,0 +1,136 @@
+	uint8_t buf[2];
+	int ret;
+
+	__ASSERT_NO_MSG(chan == SENSOR_CHAN_ALL || chan == SENSOR_CHAN_AMBIENT_TEMP);
```
Why not return ENOTSUP when channel is other than supported to this sensor instead of asserting?

maxmclau replied:
I decided to use an assertion instead after reviewing other sensor driver implementations - BH1750 is one of the few I found returning ENOTSUP instead. I thought it would be best to keep it consistent, although I don't really see the benefit.

kartben replied:
Maybe @MaureenHelm has some thoughts here. I've found many drivers use `-ENOTSUP`, which I think is useful for when one uses the sensor shell commands (i.e it automagically outputs sensor values only for the channels supported)?

uLipe commented on drivers/sensor/mcp9600/mcp9600.c:
```diff
@@ -0,0 +1,136 @@
+	data->temp = (int32_t)(int16_t)(buf[0] << 8) | buf[1];
+
+	/* 0.0625C resolution per LSB */
+	data->temp *= 62500;
```
This constant could be places into this sensor binding allowing to be set per-application.

maxmclau replied:
I don't love having magic numbers floating around either. I'd have to add support for modifying the ADC resolution though - maybe a future feature

uLipe reviewed (dismissed):
LGTM, left some suggestions,but feel free to ignore them.

carlescufi commented:
@maxmclau please rebase

maxmclau commented:
@carlescufi rebased

kartben commented on drivers/sensor/mcp9600/Kconfig:
```diff
@@ -0,0 +1,14 @@
+# SPDX-License-Identifier: Apache-2.0
+
+config MCP9600
+	bool "MCP9600 i2c thermocouple temperature sensor"
```
nit: this could probably be "I2C", not "i2c"
